pkg/vsphere/vm/vm.go

```diff
@@ -731,3 +734,149 @@ func (vm *VirtualMachine) RemoveSnapshotByRef(ctx context.Context, snapshot *typ
+func (vm *VirtualMachine) PowerOn(op trace.Operation) error {
+	drs := vm.Session.DRSEnabled != nil && *vm.Session.DRSEnabled
+	cls := vm.InCluster(op)
+
+	op.Debugf("DRS enabled: %t", drs)
+	op.Debugf("%s resides in a cluster: %t", vm.Reference.String(), cls)
+
+	// otherwise place the VM before powering it on
+	hmp := performance.NewHostMetricsProvider(vm.Session.Vim25())
+	rhp, err := placement.NewRankedHostPolicy(op, vm.Cluster, hmp)
+	if err != nil {
+		return err
+	}
+
+	// any of the following conditions are sufficient to power-on the VM:
+	// - environment is not VC
+	// - DRS is enabled
+	// - the VM does not belong to a cluster
+	// - the current host is adequate for power-on
+	if !vm.IsVC() || drs || !cls || rhp.CheckHost(op, vm.VirtualMachine) {
+		return vm.powerOn(op)
+	}
+
+	op.Warnf("Host is not adequate for power-on, getting placement recommendation")
+
+	var hosts, subset []*object.HostSystem
+
+	f := func() error {
+		hosts, err = rhp.RecommendHost(op, subset)
+		if err != nil {
+			return fmt.Errorf("error recommending host: %s", err.Error())
+		}
+		return nil
+	}
+
+	conf := retry.NewBackoffConfig()
+	conf.MaxInterval = 1 * time.Second
+	conf.MaxElapsedTime = 20 * time.Second
+
+	err = retry.DoWithConfig(f, retry.OnError, conf)
+	if err != nil {
+		return err
+	}
+
+	for len(hosts) > 0 {
+		op.Debugf("hosts: %s", hosts)
+		op.Infof("Placement recommended %s", hosts[0].Reference().String())
+
+		if vm.relocate(op, hosts[0]) != nil {
+			op.Warnf("VM relocation failed: %s", err.Error())
+		} else if vm.powerOn(op) != nil {
+			op.Warnf("VM power-on failed: %s", err.Error())
+		} else {
+			return nil
+		}
+
+		// if relocation or powerOn fails, something has changed and we need a new recommendation
+		subset = hosts[1:]
+
+		if err = retry.DoWithConfig(f, retry.OnError, conf); err != nil {
+			return err
+		}
+
+	}
+	return fmt.Errorf("vm placement failed: no available hosts")
+}
+
+func (vm *VirtualMachine) relocate(op trace.Operation, host *object.HostSystem) error {
+	h, err := vm.HostSystem(op)
+	if err != nil {
+		return err
+	}
+	src := h.Reference()
+	dst := host.Reference()
+
+	// NOP if dest host and src host are the same
+	if src.String() == dst.String() {
+		op.Debugf("Skipping relocate - source and destination hosts are the same")
+		return nil
+	}
+
+	op.Debugf("Attempting to relocate %s from host %s to host %s", vm.Reference().String(), src.String(), dst.String())
+
+	spec := types.VirtualMachineRelocateSpec{Host: &dst}
+	_, err = vm.WaitForResult(op, func(op context.Context) (tasks.Task, error) {
+		return vm.Relocate(op, spec, types.VirtualMachineMovePriorityDefaultPriority)
+	})
+	if err != nil {
+		return err
+	}
+
+	op.Infof("VM %s successfully relocated", vm.Reference().String())
+
+	return nil
+}
+
+func (vm *VirtualMachine) powerOn(op trace.Operation) error {
+	// TODO(jzt): use DRS aware power-on
```

Comment:
Do we need this commented out code?  We've got an issue open for it...so not sure we need the _dead_ code

Reply:
Nah, I'll get rid of it.